File: app/src/main/java/com/xomena/cmpfutboltfe/StreetViewRouteStepActivity.java
package com.xomena.cmpfutboltfe;
import com.xomena.cmpfutboltfe.model.*;
import com.xomena.cmpfutboltfe.util.*;

import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.res.ResourcesCompat;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.OnStreetViewPanoramaReadyCallback;
import com.google.android.gms.maps.StreetViewPanorama;
import com.google.android.gms.maps.StreetViewPanoramaFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.android.gms.maps.model.StreetViewPanoramaCamera;
import com.google.android.gms.maps.model.StreetViewPanoramaLocation;
import com.google.maps.android.PolyUtil;

import java.util.List;

public class StreetViewRouteStepActivity extends AppCompatActivity
        implements OnStreetViewPanoramaReadyCallback, OnMapReadyCallback,
        StreetViewPanorama.OnStreetViewPanoramaChangeListener {

    private static final String LOG_TAG = "StreetViewRouteStep";

    private String polyline;
    List<LatLng> path;
    private StreetViewPanorama mStreetViewPanorama;
    private LatLng posLatLng;
    private FootballField ff;
    private GoogleMap map;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_street_view_route_step);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbarSVStep);
        setSupportActionBar(toolbar);
        try {
            ActionBar ab = getSupportActionBar();
            if (ab != null) {
                ab.setDisplayShowTitleEnabled(false);
            }
        } catch (NullPointerException e) {
            Log.e(LOG_TAG, "Exception", e);
        }

        final Drawable upArrow = ResourcesCompat.getDrawable(getResources(), R.drawable.arrow_left,
                getApplicationContext().getTheme());
        toolbar.setNavigationIcon(upArrow);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });

        Intent i = getIntent();
        double m_lat = i.getDoubleExtra(MainActivity.SV_LAT, 0);
        double m_lng = i.getDoubleExtra(MainActivity.SV_LNG, 0);
        posLatLng = new LatLng(m_lat, m_lng);

        polyline = i.getStringExtra(MainActivity.EXTRA_ENC_POLY);
        if(polyline!=null){
            path = PolyUtil.decode(polyline);
        }
        ff = i.getParcelableExtra(MainActivity.EXTRA_ITEM);
        String descr = i.getStringExtra("ROUTE_STEP_DESCR");
        TextView textDescr = (TextView) findViewById(R.id.sv_step_descr);
        textDescr.setText(descr);

        MapFragment mapFrag = (MapFragment)
                getFragmentManager().findFragmentById(R.id.sv_step_map);
        mapFrag.getMapAsync(this);

        StreetViewPanoramaFragment streetViewPanoramaFragment =
                (StreetViewPanoramaFragment) getFragmentManager()
                        .findFragmentById(R.id.sv_step_panorama);
        streetViewPanoramaFragment.getStreetViewPanoramaAsync(this);
    }

    @Override
    public void onStreetViewPanoramaReady(StreetViewPanorama panorama) {
        mStreetViewPanorama = panorama;
        panorama.setOnStreetViewPanoramaChangeListener(this);
        panorama.setPosition(posLatLng, 5);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        map = googleMap;
        googleMap.getUiSettings().setMapToolbarEnabled(false);
        if(polyline != null && path!=null && path.size()>0) {
            PolylineOptions polyOptions = new PolylineOptions().addAll(path);
            googleMap.addPolyline(polyOptions);

            googleMap.addMarker(new MarkerOptions()
                    .position(posLatLng).anchor(0.5f, 0.5f).draggable(false)
                    .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_point)));

            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    LatLngBounds.Builder builder = new LatLngBounds.Builder();
                    for(LatLng coord : path){
                        builder.include(coord);
                    }
                    LatLngBounds m_bounds = builder.build();

                    map.moveCamera(CameraUpdateFactory.newLatLngBounds(m_bounds, 10));
                }
            }, 1000);
        }
    }

    @Override
    public void onStreetViewPanoramaChange(StreetViewPanoramaLocation location) {
        LatLng nearest = null;
        int nearestInd = -1;
        float mdist = 100000f;
        int index = -1;
        for(LatLng rpoint : path) {
            index++;
            float d = distanceBetween(posLatLng, rpoint);
            if (d < mdist) {
                mdist = d;
                nearest = rpoint;
                nearestInd = index;
            }
        }
        if (nearest != null && nearestInd != -1) {
            LatLng nextLatLng;
            if (nearestInd + 1 < path.size()-1) {
                nextLatLng = path.get(nearestInd + 1);
            } else {
                nextLatLng = new LatLng(ff.getLat(), ff.getLng());
            }
            //Find the Bearing from current location to next location
            float targetBearing = getTargetBearing(nextLatLng);

            long duration = 100;
            mStreetViewPanorama.animateTo(new StreetViewPanoramaCamera.Builder().
                    bearing(targetBearing).build(),duration);
        }
    }

    private float distanceBetween(LatLng latLng1, LatLng latLng2) {

        Location loc1 = new Location(LocationManager.GPS_PROVIDER);
        Location loc2 = new Location(LocationManager.GPS_PROVIDER);

        loc1.setLatitude(latLng1.latitude);
        loc1.setLongitude(latLng1.longitude);

        loc2.setLatitude(latLng2.latitude);
        loc2.setLongitude(latLng2.longitude);


        return loc1.distanceTo(loc2);
    }

    //Gets bearing between current point and next point in path
    private float getTargetBearing(LatLng next) {
        float targetBearing = 0f;
        if (next != null) {
            Location startLocation = new Location("starting point");
            Location endLocation = new Location("ending point");
            startLocation.setLatitude(posLatLng.latitude);
            startLocation.setLongitude(posLatLng.longitude);
            endLocation.setLatitude(next.latitude);
            endLocation.setLongitude(next.longitude);
            //Find the Bearing from current location to next location
            targetBearing = startLocation.bearingTo(endLocation);
        }
        return targetBearing;
    }

}

File: app/src/main/java/com/xomena/cmpfutboltfe/StreetViewRouteStepActivity.kt
package com.xomena.cmpfutboltfe

import com.xomena.cmpfutboltfe.model.*

import android.content.Intent
import android.graphics.drawable.Drawable
import android.location.Location
import android.location.LocationManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.TextView

import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.res.ResourcesCompat

import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapFragment
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.OnStreetViewPanoramaReadyCallback
import com.google.android.gms.maps.StreetViewPanorama
import com.google.android.gms.maps.StreetViewPanoramaFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.PolylineOptions
import com.google.android.gms.maps.model.StreetViewPanoramaCamera
import com.google.android.gms.maps.model.StreetViewPanoramaLocation
import com.google.maps.android.PolyUtil

class StreetViewRouteStepActivity : AppCompatActivity(),
    OnStreetViewPanoramaReadyCallback, OnMapReadyCallback,
    StreetViewPanorama.OnStreetViewPanoramaChangeListener {

    private var polyline: String? = null
    private var path: List<LatLng>? = null
    private var mStreetViewPanorama: StreetViewPanorama? = null
    private var posLatLng: LatLng? = null
    private var ff: FootballField? = null
    private var map: GoogleMap? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_street_view_route_step)

        val toolbar = findViewById<Toolbar>(R.id.toolbarSVStep)
        setSupportActionBar(toolbar)
        try {
            val ab = supportActionBar
            ab?.setDisplayShowTitleEnabled(false)
        } catch (e: NullPointerException) {
            Log.e(LOG_TAG, "Exception", e)
        }

        val upArrow = ResourcesCompat.getDrawable(
            resources, R.drawable.arrow_left,
            applicationContext.theme
        )
        toolbar.navigationIcon = upArrow
        toolbar.setNavigationOnClickListener { onBackPressed() }

        val i = intent
        val m_lat = i.getDoubleExtra(MainActivity.SV_LAT, 0.0)
        val m_lng = i.getDoubleExtra(MainActivity.SV_LNG, 0.0)
        posLatLng = LatLng(m_lat, m_lng)

        polyline = i.getStringExtra(MainActivity.EXTRA_ENC_POLY)
        polyline?.let {
            path = PolyUtil.decode(it)
        }
        ff = i.getParcelableExtra(MainActivity.EXTRA_ITEM)
        val descr = i.getStringExtra("ROUTE_STEP_DESCR")
        val textDescr = findViewById<TextView>(R.id.sv_step_descr)
        textDescr.text = descr

        val mapFrag = fragmentManager.findFragmentById(R.id.sv_step_map) as MapFragment
        mapFrag.getMapAsync(this)

        val streetViewPanoramaFragment =
            fragmentManager.findFragmentById(R.id.sv_step_panorama) as StreetViewPanoramaFragment
        streetViewPanoramaFragment.getStreetViewPanoramaAsync(this)
    }

    override fun onStreetViewPanoramaReady(panorama: StreetViewPanorama) {
        mStreetViewPanorama = panorama
        panorama.setOnStreetViewPanoramaChangeListener(this)
        posLatLng?.let {
            panorama.setPosition(it, 5)
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap
        googleMap.uiSettings.isMapToolbarEnabled = false
        val currentPath = path
        if (polyline != null && currentPath != null && currentPath.isNotEmpty()) {
            val polyOptions = PolylineOptions().addAll(currentPath)
            googleMap.addPolyline(polyOptions)

            posLatLng?.let {
                googleMap.addMarker(
                    MarkerOptions()
                        .position(it).anchor(0.5f, 0.5f).draggable(false)
                        .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_point))
                )
            }

            val handler = Handler(Looper.getMainLooper())
            handler.postDelayed({
                val builder = LatLngBounds.Builder()
                for (coord in currentPath) {
                    builder.include(coord)
                }
                val m_bounds = builder.build()

                map?.moveCamera(CameraUpdateFactory.newLatLngBounds(m_bounds, 10))
            }, 1000)
        }
    }

    override fun onStreetViewPanoramaChange(location: StreetViewPanoramaLocation) {
        var nearest: LatLng? = null
        var nearestInd = -1
        var mdist = 100000f
        var index = -1
        
        val currentPos = posLatLng ?: return
        
        for (rpoint in path ?: emptyList()) {
            index++
            val d = distanceBetween(currentPos, rpoint)
            if (d < mdist) {
                mdist = d
                nearest = rpoint
                nearestInd = index
            }
        }
        if (nearest != null && nearestInd != -1) {
            val nextLatLng: LatLng = if (nearestInd + 1 < (path?.size ?: 0) - 1) {
                path!![nearestInd + 1]
            } else {
                LatLng(ff?.lat ?: 0.0, ff?.lng ?: 0.0)
            }
            //Find the Bearing from current location to next location
            val targetBearing = getTargetBearing(nextLatLng)

            val duration: Long = 100
            mStreetViewPanorama?.animateTo(
                StreetViewPanoramaCamera.Builder()
                    .bearing(targetBearing).build(), duration
            )
        }
    }

    private fun distanceBetween(latLng1: LatLng, latLng2: LatLng): Float {
        val loc1 = Location(LocationManager.GPS_PROVIDER)
        val loc2 = Location(LocationManager.GPS_PROVIDER)

        loc1.latitude = latLng1.latitude
        loc1.longitude = latLng1.longitude

        loc2.latitude = latLng2.latitude
        loc2.longitude = latLng2.longitude

        return loc1.distanceTo(loc2)
    }

    //Gets bearing between current point and next point in path
    private fun getTargetBearing(next: LatLng?): Float {
        var targetBearing = 0f
        val currentPos = posLatLng
        if (next != null && currentPos != null) {
            val startLocation = Location("starting point")
            val endLocation = Location("ending point")
            startLocation.latitude = currentPos.latitude
            startLocation.longitude = currentPos.longitude
            endLocation.latitude = next.latitude
            endLocation.longitude = next.longitude
            //Find the Bearing from current location to next location
            targetBearing = startLocation.bearingTo(endLocation)
        }
        return targetBearing
    }

    companion object {
        private const val LOG_TAG = "StreetViewRouteStep"
    }
}
